aziot/src/internal/check/checks/mod.rs

```diff
@@ -0,0 +1,96 @@
+    };
+
+    pub trait CertificateValidityExt {
+        fn to_check_result(&self) -> Result<CheckResult>;
+    }
+
+    impl CertificateValidityExt for crate::internal::common::CertificateValidity {
+        fn to_check_result(&self) -> Result<CheckResult> {
+            let now = chrono::Utc::now();
+            if self.not_before > now {
+                Err(anyhow!(
+                    "{} '{}' has not-before time {} which is in the future",
+                    self.cert_name,
+                    self.cert_id,
+                    self.not_before,
+                ))
+            } else if self.not_after < now {
+                Err(anyhow!(
+                    "{} '{}' expired at {}",
+                    self.cert_name,
+                    self.cert_id,
+                    self.not_after,
+                ))
+            } else if self.not_after < now + chrono::Duration::days(7) {
+                Ok(CheckResult::Warning(anyhow!(
+                    "{} '{}' will expire soon ({}, in {} days)",
+                    self.cert_name,
+                    self.cert_id,
+                    self.not_after,
+                    (self.not_after - now).num_days(),
+                )))
+            } else {
+                Ok(CheckResult::Ok)
+            }
+        }
+    }
+}
+
+/// Tries to unwrap an option, early-returning with
+/// `return Ok(CheckResult::Skipped)` if the option is None.
+macro_rules! unwrap_or_skip {
+    ($opt:expr) => {{
+        use crate::internal::check::CheckResult;
+
+        match $opt {
+            Some(val) => val,
+            None => return Ok(CheckResult::Skipped),
+        }
+    }};
+}
+
+mod certs_preloaded;
+mod daemons_running;
+mod host_connect_dps_endpoint;
+mod host_local_time;
+mod hostname;
+mod identity_certificate_expiry;
+mod well_formed_configs;
+
+pub fn all_checks() -> Vec<(&'static str, Vec<Box<dyn Checker>>)> {
+    // DEVNOTE: keep ordering consistent. Later tests may depend on earlier tests.
+    vec![
+        ("Configuration checks", {
+            let mut v: Vec<Box<dyn Checker>> = Vec::new();
+            v.extend(well_formed_configs::well_formed_configs());
+            v.push(Box::new(hostname::Hostname::default()));
+            // TODO: add aziot version info to https://github.com/Azure/azure-iotedge
+            // v.push(Box::new(aziot_version::AziotVersion::default()));
+            v.push(Box::new(host_local_time::HostLocalTime::default()));
+            v.push(Box::new(
+                identity_certificate_expiry::IdentityCertificateExpiry::default(),
+            ));
+            v.push(Box::new(certs_preloaded::CertsPreloaded::default()));
+            v
+        }),
+        ("Connectivity checks", {
+            let mut v: Vec<Box<dyn Checker>> = Vec::new();
+            v.extend(daemons_running::daemons_running());
+            v.push(Box::new(
+                host_connect_dps_endpoint::HostConnectDpsEndpoint::default(),
```

Comment:
Is iot hub connectivity checks in a different PR?

Reply:
Are you talking about `host_connect_iothub`?

I was going to leave those in `iotedge check`, since fixing them is as easy as passing the `iothub_hostname` up from `aziot check` to `iotedge check` (through the `AdditionalInfo` payload).

That said, reviewing the [check's implementation](https://github.com/Azure/iotedge/blob/release/1.0.10/edgelet/iotedge/src/check/checks/host_connect_iothub.rs), there's no technical reason why it couldn't be pushed down into `aziot check`. Then again, does it make sense to perform AMQP and MQTT connectivity checks as part of the `aziot check`? My gut says no, not really.

Let me know what you think.

Reply:
>I was going to leave those in iotedge check, since fixing them is as easy as passing the iothub_hostname up from aziot check to iotedge check (through the AdditionalInfo payload).

We cannot assume `iotedge` exists on the same machine as `aziot`. The TCP connectivity check would be the same on both sides, but the protocols are different (HTTPS in `aziot check` vs AMQP/MQTT (+ WS) in `iotedge check`).

Reply:
Sounds good. In that case, I'll go ahead and add the HTTPS check into `aziot check`.

Reply:
We do need AMQP/MQTT (+ WS) in `aziot check` too.

Reply:
Why?

Reply:
The same reason `iotedge check` has them, even though `iotedged` doesn't use them either.

The point of `check` is to not just check the protocols our code uses, but what the user's code might use. That means host modules, which would use those protocols.

Reply:
But doesn't that reinforce the point @massand made? i.e: if the check is there to support a user's code, then shouldn't [some of] the the checks stay in iotedge?

An alternative argument might be that having two nearly identical checks targeting different protocols is a bit wasteful, and that we should just consolidate everything in either `aziot check` or `iotedge check` - likely the former.

Reply:
I don't understand the confusion.

Host-connectivity checks are needed to troubleshoot host modules. Host modules are serviced by IS. The presence or absence of `iotedged` is irrelevant. Conclusion: `aziot check` needs host-to-upstream connectivity checks.

Container-connectivity checks are needed to troubleshoot Edge modules. Edge modules are serviced by `iotedged`. Conclusion: `iotedge check` needs container-to-upstream connectivity checks. Since it also delegates to `aziot check`, `iotedge check` does not need host-to-upstream connectivity checks.

Reply:
Oh, shoot, yes, that's right. I forgot that host modules are a thing (or at least, will be more of a thing).

Cool, in that case, I'll migrate the entire check into `aziot check`.

Reply:
Ooooh, now I remember why I wanted to leave it in `iotedge check`: the check doesn't has additional logic to ping the `parent_hostname` in nested-edge scenarios: https://github.com/Azure/iotedge/blob/release/1.2/edgelet/iotedge/src/check/checks/host_connect_upstream.rs#L66-L74

In any case, I have gone ahead and added a iothub connectivity check to `aziot check`, but what should we do about the nested-edge stuff?

Reply:
As I said in the other comment, the `parent_hostname` check is part of the IS already via `aziot check` in the iothub connectivity check. I think if you add a check in `iotedge check` for `parent_hostname` matching the `iothub_hostname` in the IS. BTW, to get the `iothub_hostname`, I would read the `prev_state` file in the IS home folder, instead of the IS `config.toml` because that would be the single source of truth when DPS provisioning is supported for Nested Edge (currently Nested Edge only supports manual provisioning).

Reply:
Ahh, interesting... I like that approach a lot. 

As for how to obtain `iothub_hostname` - I'm currently including it as part of the `aziot check` `additional_info` JSON payload, which gets consumed by `iotedge` when it shells-out to `aziot check`. Let me know if you forsee any issues with that approach, though I think it's a pretty clean way to solving the "single source of truth" problem.